private/reducer.rkt

```diff
@@ -69,61 +71,60 @@
 (define-reference-type reducer (starter consumer finisher early-finisher))
 
 (define (make-fold-reducer consumer init-state #:name [name #f])
-  (make-effectful-fold-reducer consumer (λ () init-state) values #:name name))
+  (make-effectful-fold-reducer consumer (fn init-state) values #:name name))
 
 (define (make-effectful-fold-reducer consumer init-state-maker finisher
                                      #:name [name #f])
-  (make-reducer #:starter (λ () (variant #:consume (init-state-maker)))
-                #:consumer (λ (s v) (variant #:consume (consumer s v)))
+  (make-reducer #:starter (fn #/variant #:consume #/init-state-maker)
+                #:consumer (fn s v #/variant #:consume #/consumer s v)
                 #:finisher finisher
                 #:early-finisher values
                 #:name name))
 
 (define into-sum (make-fold-reducer + 0 #:name 'into-sum))
 (define into-product (make-fold-reducer * 1 #:name 'into-product))
-(define into-count (make-fold-reducer (λ (s _) (add1 s)) 0 #:name 'into-count))
+(define into-count (make-fold-reducer (fn s v #/add1 s) 0 #:name 'into-count))
```

Comment:
I use underscores for unused variables, does `fn` support that?

Reply:
I'm glad you brought this up. While `(fn s _ #/add1 s)` would correspond better with `(λ (s _) (add1 s))`, and you can easily use that if you prefer, I like to use `_` as a match pattern, and local variables with the same name can shadow it.

I'd like to use `_` to ignore arguments in a `fn`, and you can see elsewhere I do use `(dissectfn _ ...)` for unary functions that ignore their argument. If I allow match patterns in `fn` in their full generality, they could cause run time errors when they fail, and I tend to prefer run time errors to be explicitly marked with certain words like "dissect" or "err" or "force" or "assert" or "verify," so I feel like this would be more of a feature for `dissectfn` than for `fn` itself.

Even then, there's a reason I haven't generalized `dissectfn` to more than one argument. In the context of the Racket ecosystem, it could be confusing when someone writes more than one expression in a `dissectfn` body (expecting behavior like Racket `lambda`) and all but the last ends up being treated as a compound pattern instead.

Even for my own purposes, where I'm using Racket primarily for its resemblance to Cene and hence try to conform to Cene-based expectations more than Racket-based ones, I find it awfully rare that I need to need to write more than one term in a function body. After all, Cene doesn't have ambient access to side effects aside from errors and nontermination, and I didn't think Cene would necessarily have an internal definition system until recently. So when I do end up using one or the other because Racket makes them easy to use, I mark this special occasion with an explicit `(begin ...)` or `(let () ...)`. So when I write too many expressions in a `dissectfn` or a `fn` or an `expect` -- and sometimes I do indeed -- I like for it to be an error.